src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
+            head = newNode;
+        } else if (index == size) {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> current = getNodeByIndex(index);
+            newNode.next = current;
+            newNode.prev = current.prev;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        return (T) current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        outOfBoundsCheck(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfBoundsCheck(index);
+        if (index == 0) {
+            T removedElement = (T) head.value;
+            head = head.next;
+            size--;
+            return removedElement;
+        } else if (index == size - 1) {
+            T removedElement = (T) tail.value;
+            tail = tail.prev;
+            size--;
+            return removedElement;
+        }
+        T removedElement;
+        Node<T> current = getNodeByIndex(index);
+        removedElement = (T) current.value;
+        unlink(current);
+        size--;
+        return removedElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int i = 0; i < size; i++) {
+            if (current.value != null && current.value.equals(object) || current.value == object) {
+                unlink(current);
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void outOfBoundsCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Wrong index: " + index);
+        }
+    }
+
+    private void checkAddedIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException("Wrong index: " + index);
+        }
+    }
+
+    private void unlink(Node current) {
+        if (current == head) {
+            head = head.next;
+        } else if (current == tail) {
+            tail = tail.prev;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
```

Comment:
I recommend using the following implementation of the method:
```
 private Node<T> unlink(Node<T> current) {
        if (current.prev == null) {
            head = current.next;
        } else {
            current.prev.next = current.next;
        }
        if (current.next == null) {
            tail = current.prev;
        } else {
            current.next.prev = current.prev;
        }
        size--;
        return current;
    }
    
```
In this implementation, we return the node and also automatically reduce the size directly encapsulated in this method